core/src/consensus.rs

```diff
@@ -515,6 +515,60 @@ impl Tower {
             .map(|last_voted_slot| {
                 let root = self.root();
                 let empty_ancestors = HashSet::default();
+                let empty_ancestors_due_to_minor_unsynced_ledger = || {
+                    // This condition (stale stray last vote) shouldn't occur under normal validator
+                    // operation, indicating something unusual happened.
+                    // This condition could be introduced by manual ledger mishandling,
+                    // validator SEGV, OS/HW crash, or plain No Free Space FS error.
+
+                    // However, returning empty ancestors as a fallback here shouldn't result in
+                    // slashing by itself (Note that we couldn't fully preclude any kind of slashing if
+                    // the failure was OS or HW level).
+
+                    // Firstly, lockout is ensured elsewhere.
+
+                    // Also, there is no risk of optimistic conf. violation. Although empty ancestors
+                    // could result in incorrect (= more than actual) locked_out_stake and
+                    // false-positive SwitchProof later in this function, there should be no such a
+                    // heavier fork candidate, first of all, if the last vote (or any of its
+                    // unavailable ancestors) were already optimistically confirmed.
+                    // The only exception is that other validator is already violating it...
+                    if self.is_first_switch_check() && switch_slot < last_voted_slot {
+                        // `switch < last` is needed not to warn! this message just because of using
+                        // newer snapshots on validator restart
+                        let message = format!(
+                          "bank_forks doesn't have corresponding data for the stray restored \
+                           last vote({}), meaning some inconsistency between saved tower and ledger.",
+                          last_voted_slot
+                        );
+                        warn!("{}", message);
+                        datapoint_warn!("tower_warn", ("warn", message, String));
+                    }
+                    &empty_ancestors
+                };
+
+                let suspended_decision_due_to_major_unsynced_ledger = |total_stake: u64| -> SwitchForkDecision {
+                    // This peculiar corner handling is needed mainly for a tower which is newer than
+                    // blockstore. (Yeah, we tolerate it for ease of maintaining validator by operators)
+                    // This condition could be introduced by manual ledger mishandling,
+                    // validator SEGV, OS/HW crash, or plain No Free Space FS error.
+
+                    // When we're in this clause, it basically means validator is badly running
+                    // with a future tower and RE-CREATING a block for one of past slots as the leader
```

Comment:
hmm, the validator doesn't even have to be leader/recreating a block for this to happen right? For example in this example here: https://github.com/solana-labs/solana/pull/12671#discussion_r513830724, validator `A` doesn't necessarily have to have generated blocks 1-4, those could be any validator's blocks, the key is it will still try to call the switching logic on a previously seen block.

Side note, could we include a simple diagram like the one in https://github.com/solana-labs/solana/pull/12671#discussion_r513830724 in the comment? Helps to clarify these points 😃

Reply:
> the validator doesn't even have to be leader/recreating a block for this to happen right?

You're correct. it seems validator doesn't have to be leader. I got false impression of leader requirement. Because, unstaked validators seldom reached to this closure while testing locally and manually.

Comment updated: 677887aafabc819ebbebd7959239648af8ce4e78

Now, I improved test_future_tower to test for leader and non-leader to validate this: 7d60a0b89b13c8e8af0e4f9f9baf22efd6bc0c10

Reply:
> Side note, could we include a simple diagram like the one in [#12671 (comment)](https://github.com/solana-labs/solana/pull/12671#discussion_r513830724) in the comment? Helps to clarify these points smiley

I'll do this later: https://github.com/solana-labs/solana/pull/12671#issuecomment-719473558